File: src/main/scala/dev/alexnader/framity/client/FrameTransform.scala
package dev.alexnader.framity.client

import java.util.Random
import java.util.function.Supplier
import java.{util => ju}

import dev.alexnader.framity.block.frame.{FrameAccess, FrameBlock}
import dev.alexnader.framity.client.assets.overlay.{IdentifierOverlayAccess, OverlayInfo, TextureSource}
import dev.alexnader.framity.util.ScalaExtensions.{ToTuple, Tuple4Ext}
import grondag.jmx.api.QuadTransformRegistry.QuadTransformSource
import net.fabricmc.api.{EnvType, Environment}
import net.fabricmc.fabric.api.client.rendering.v1.ColorProviderRegistry
import net.fabricmc.fabric.api.renderer.v1.RendererAccess
import net.fabricmc.fabric.api.renderer.v1.mesh.MutableQuadView
import net.fabricmc.fabric.api.renderer.v1.render.RenderContext.QuadTransform
import net.fabricmc.fabric.api.rendering.data.v1.RenderAttachedBlockView
import net.minecraft.block.BlockState
import net.minecraft.client.MinecraftClient
import net.minecraft.client.texture.{Sprite, SpriteAtlasTexture}
import net.minecraft.item.ItemStack
import net.minecraft.util.Identifier
import net.minecraft.util.math.{BlockPos, Direction, MathHelper, Vec3d}
import net.minecraft.world.BlockRenderView
import grondag.frex.api.{Renderer => FrexRenderer}
import grondag.frex.api.material.{RenderMaterial => FrexRenderMaterial}

import scala.collection.View

@Environment(EnvType.CLIENT)
object FrameTransform {
  import dev.alexnader.framity.client.util.SpriteSource

  private def getUvs(mqv: MutableQuadView, direction: Direction): ((Float, Float, Float, Float), (Float, Float, Float, Float)) = direction match {
    case Direction.DOWN => (
      (0 to 3).map(i => MathHelper.clamp(mqv.x(i), 0f, 1f)).toTuple4,
      (0 to 3).map(i => 1 - MathHelper.clamp(mqv.z(i), 0f, 1f)).toTuple4
    )
    case Direction.UP => (
      (0 to 3).map(i => MathHelper.clamp(mqv.x(i), 0f, 1f)).toTuple4,
      (0 to 3).map(i => MathHelper.clamp(mqv.z(i), 0f, 1f)).toTuple4
    )
    case Direction.NORTH => (
      (0 to 3).map(i => 1 - MathHelper.clamp(mqv.x(i), 0f, 1f)).toTuple4,
      (0 to 3).map(i => 1 - MathHelper.clamp(mqv.y(i), 0f, 1f)).toTuple4
    )
    case Direction.SOUTH => (
      (0 to 3).map(i => MathHelper.clamp(mqv.x(i), 0f, 1f)).toTuple4,
      (0 to 3).map(i => 1 - MathHelper.clamp(mqv.y(i), 0f, 1f)).toTuple4
    )
    case Direction.EAST => (
      (0 to 3).map(i => 1 - MathHelper.clamp(mqv.z(i), 0f, 1f)).toTuple4,
      (0 to 3).map(i => 1 - MathHelper.clamp(mqv.y(i), 0f, 1f)).toTuple4
    )
    case Direction.WEST => (
      (0 to 3).map(i => MathHelper.clamp(mqv.z(i), 0f, 1f)).toTuple4,
      (0 to 3).map(i => 1 - MathHelper.clamp(mqv.y(i), 0f, 1f)).toTuple4
    )
  }

  private def getSprite(direction: Direction, source: TextureSource): Option[Sprite] = source match {
    case TextureSource.Sided(map) => map.get(direction).flatMap(id => Option(MinecraftClient.getInstance.getSpriteAtlas(SpriteAtlasTexture.BLOCK_ATLAS_TEX).apply(id)))
    case TextureSource.Single(id) => Option(MinecraftClient.getInstance.getSpriteAtlas(SpriteAtlasTexture.BLOCK_ATLAS_TEX).apply(id))
  }

  private def applySpriteAndColor(mqv: MutableQuadView, maybeSpriteAndColor: Option[(Sprite, Option[Int])], us: (Float, Float, Float, Float), vs: (Float, Float, Float, Float), spriteIndex: Int): Boolean = {
    maybeSpriteAndColor.map { case (sprite, maybeColor) =>
      mqv.sprite(0, spriteIndex, MathHelper.lerp(us._1, sprite.getMinU, sprite.getMaxU), MathHelper.lerp(vs._1, sprite.getMinV, sprite.getMaxV))
      mqv.sprite(1, spriteIndex, MathHelper.lerp(us._2, sprite.getMinU, sprite.getMaxU), MathHelper.lerp(vs._2, sprite.getMinV, sprite.getMaxV))
      mqv.sprite(2, spriteIndex, MathHelper.lerp(us._3, sprite.getMinU, sprite.getMaxU), MathHelper.lerp(vs._3, sprite.getMinV, sprite.getMaxV))
      mqv.sprite(3, spriteIndex, MathHelper.lerp(us._4, sprite.getMinU, sprite.getMaxU), MathHelper.lerp(vs._4, sprite.getMinV, sprite.getMaxV))

      maybeColor.tapEach { color =>
        mqv.spriteColor(spriteIndex, color, color, color, color)
      }
    }.isDefined
  }

  protected case class Data(maybeSprites: Option[SpriteSource], maybeOverlay: Option[OverlayInfo], maybeCachedOverlayColor: Option[Int], color: Int)

  object NonFrex {
    object Source extends QuadTransformSource {
      override def getForBlock(blockRenderView: BlockRenderView, blockState: BlockState, blockPos: BlockPos, supplier: Supplier[Random]): QuadTransform =
        new NonFrex(blockRenderView, blockState, blockPos, supplier)

      override def getForItem(itemStack: ItemStack, supplier: Supplier[Random]): QuadTransform = null
    }
  }

  class NonFrex(blockView: BlockRenderView, state: BlockState, pos: BlockPos, randomSupplier: Supplier[ju.Random]) extends FrameTransform(blockView, state, pos, randomSupplier) {
    private val transformedIndex: ju.EnumMap[Direction, Int] = new ju.EnumMap(classOf[Direction])

    override def transform(mqv: MutableQuadView): Boolean = {
      if (mqv.tag() == 1) {
        return true
      }

      val direction = mqv.lightFace()

      val quadIndex = transformedIndex.getOrDefault(direction, 0)
      transformedIndex.put(direction, quadIndex + 1)

      val partIndex = getPartIndex(mqv, direction)

      val Data(maybeSprites, maybeOverlay, maybeCachedOverlayColor, color) = transformData(partIndex)

      val (maybeSpriteAndColor, us, vs) = {
        val (origUs, origVs) = getUvs(mqv, direction)

        def findMaybeSpriteAndColor(sprites: SpriteSource): Option[(Sprite, Option[Int])] = {
          def cyclicQuadIndex(sprites: SpriteSource): Int = quadIndex % sprites.getCount(direction).getOrElse(1)

          sprites(direction, cyclicQuadIndex(sprites), color)
        }

        if (quadIndex % 2 == 0) { // base quads are first, overlay quads are second
          maybeSprites match {
            case None => return true // don't transform if this quad is base and there's nothing to apply
            case Some(sprites) =>
              (findMaybeSpriteAndColor(sprites), origUs, origVs)
          }
        } else {
          val maybeSpriteAndColor = maybeOverlay match {
            case None => maybeSprites.flatMap(findMaybeSpriteAndColor) // if no overlay, check if base model has more than 1 quad per face and use that
            case Some(overlay) =>
              val maybeSprite = overlay.textureSource.map(textureSource => getSprite(direction, textureSource)) match {
                case Some(None) => return false // there is a texture to use but it doesn't have anything on this direction
                case Some(Some(sprite)) => Some(sprite)
                case None => None
              }

              maybeSprite.map(sprite => (sprite, maybeCachedOverlayColor))
          }

          val (us, vs) = maybeOverlay
            .flatMap(_.getUvs(direction, origUs, origVs))
            .getOrElse(origUs, origVs)

          (maybeSpriteAndColor, us, vs)
        }
      }

      applySpriteAndColor(mqv, maybeSpriteAndColor, us, vs, 0)

      maybeSpriteAndColor.isDefined
    }
  }

  object Frex {
    object Source extends QuadTransformSource {
      override def getForBlock(blockRenderView: BlockRenderView, blockState: BlockState, blockPos: BlockPos, supplier: Supplier[Random]): QuadTransform =
        new Frex(blockRenderView, blockState, blockPos, supplier)

      override def getForItem(itemStack: ItemStack, supplier: Supplier[Random]): QuadTransform = null
    }
  }

  class Frex(blockView: BlockRenderView, state: BlockState, pos: BlockPos, randomSupplier: Supplier[ju.Random]) extends FrameTransform(blockView, state, pos, randomSupplier) {
    override def transform(mqv: MutableQuadView): Boolean = {
      val direction = mqv.lightFace
      val partIndex = getPartIndex(mqv, direction)
      val Data(maybeSprites, maybeOverlay, maybeCachedOverlayColor, color) = transformData(partIndex)
      val (origUs, origVs) = getUvs(mqv, direction)

      //#region Base
      maybeSprites.tapEach { sprites =>
        applySpriteAndColor(mqv, sprites(direction, 0, color), origUs, origVs, 0)
      }
      //#endregion

      //#region Overlay
      val (maybeOverlaySpriteAndColor, us, vs) = maybeOverlay match {
        case Some(overlay) =>
          val maybeOverlaySpriteAndColor = {
            val maybeSprite = overlay.textureSource
              .flatMap { getSprite(direction, _) }

            if (maybeSprite.isEmpty) {
              mqv.material(RendererAccess.INSTANCE.getRenderer.asInstanceOf[FrexRenderer].materialFinder.copyFrom(mqv.material().asInstanceOf[FrexRenderMaterial]).spriteDepth(1).find())
            }

            maybeSprite.map((_, maybeCachedOverlayColor))
          }

          val (us, vs) = overlay.getUvs(direction, origUs, origVs).getOrElse((origUs, origVs))

          (maybeOverlaySpriteAndColor, us, vs)
        case None =>
          val maybeOverlaySpriteAndColor = maybeSprites
            .flatMap { _(direction, 1, color) }

          if (maybeOverlaySpriteAndColor.isEmpty) {
            mqv.material(RendererAccess.INSTANCE.getRenderer.asInstanceOf[FrexRenderer].materialFinder.copyFrom(mqv.material().asInstanceOf[FrexRenderMaterial]).spriteDepth(1).find())
          }

          (maybeOverlaySpriteAndColor, origUs, origVs)
      }
      applySpriteAndColor(mqv, maybeOverlaySpriteAndColor, us, vs, 1)
      //#endregion

      true
    }
  }
}

sealed abstract class FrameTransform(blockView: BlockRenderView, state: BlockState, pos: BlockPos, randomSupplier: Supplier[ju.Random]) extends QuadTransform {
  import dev.alexnader.framity.client.util.SpriteSource

  if (!state.getBlock.isInstanceOf[FrameBlock]) {
    throw new IllegalArgumentException(s"Cannot apply frame transform to non-frame block ${state.getBlock} (at $pos).")
  }

  protected val transformData: Array[FrameTransform.Data] = {
    val attachment = {
      val attachment = blockView.asInstanceOf[RenderAttachedBlockView].getBlockEntityRenderAttachment(pos)
      attachment.asInstanceOf[View[(Option[BlockState], Option[Identifier])]]
    }

    attachment.map { case (baseState, overlayId) =>
      val (color, sprites) = baseState match {
        case Some(baseState) => (
          Option(ColorProviderRegistry.BLOCK.get(baseState.getBlock)).map(_.getColor(baseState, blockView, pos, 1) | 0xFF000000).getOrElse(0xFFFFFFFF),
          Some(new SpriteSource(baseState, MinecraftClient.getInstance.getBlockRenderManager.getModel(baseState), randomSupplier.get)),
        )
        case None => (0xFFFFFFFF, None)
      }
      val overlay = overlayId.flatMap(_.getOverlay)
      val cachedOverlayColor: Option[Int] = overlay
        .flatMap(_.coloredLike)
        .flatMap(coloredLike =>
          Option(ColorProviderRegistry.BLOCK.get(coloredLike.colorSource.getBlock))
            .map(_.getColor(coloredLike.colorSource, blockView, pos, 1))
            .map(_ | 0xFF000000)
        )
      FrameTransform.Data(sprites, overlay, cachedOverlayColor, color)
    }.toArray
  }

  protected def getPartIndex(mqv: MutableQuadView, direction: Direction): Int = {
    state.getBlock.asInstanceOf[FrameAccess].getRelativeSlotAt(
      state,
      new Vec3d(
        (0 to 3).map(mqv.x).toTuple4.center,
        (0 to 3).map(mqv.y).toTuple4.center,
        (0 to 3).map(mqv.z).toTuple4.center
      ),
      direction
    )
  }
}

File: src/main/java/dev/alexnader/framity2/client/FrameTransform.java
package dev.alexnader.framity2.client;

import com.mojang.datafixers.util.Pair;
import dev.alexnader.framity2.block.FrameSlotInfo;
import dev.alexnader.framity2.block.frame.Frame;
import dev.alexnader.framity2.client.assets.ColorApplier;
import dev.alexnader.framity2.client.assets.overlay.Overlay;
import dev.alexnader.framity2.util.Float4;
import grondag.frex.api.material.MaterialMap;
import grondag.jmx.api.QuadTransformRegistry;
import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.fabricmc.fabric.api.client.rendering.v1.ColorProviderRegistry;
import net.fabricmc.fabric.api.renderer.v1.material.RenderMaterial;
import net.fabricmc.fabric.api.renderer.v1.mesh.MutableQuadView;
import net.fabricmc.fabric.api.renderer.v1.render.RenderContext;
import net.fabricmc.fabric.api.rendering.data.v1.RenderAttachedBlockView;
import net.minecraft.block.BlockState;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.texture.Sprite;
import net.minecraft.item.ItemStack;
import net.minecraft.util.Identifier;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Direction;
import net.minecraft.util.math.MathHelper;
import net.minecraft.util.math.Vec3d;
import net.minecraft.world.BlockRenderView;

import javax.annotation.Nullable;
import java.util.*;
import java.util.function.Function;
import java.util.function.Supplier;
import java.util.stream.IntStream;
import java.util.stream.Stream;

import static dev.alexnader.framity2.client.Framity2Client.CLIENT_OVERLAYS;
import static dev.alexnader.framity2.client.util.QuadUtil.calcCenter;
import static dev.alexnader.framity2.util.FunctionalUtil.*;

@Environment(EnvType.CLIENT)
public final class FrameTransform implements RenderContext.QuadTransform {
    public static final QuadTransformRegistry.QuadTransformSource SOURCE = new QuadTransformRegistry.QuadTransformSource() {
        @Override
        public RenderContext.QuadTransform getForBlock(final BlockRenderView brv, final BlockState state, final BlockPos pos, final Supplier<Random> randomSupplier) {
            return new FrameTransform(brv, state, pos, randomSupplier);
        }

        @Nullable
        @Override
        public RenderContext.QuadTransform getForItem(final ItemStack itemStack, final Supplier<Random> supplier) {
            return null;
        }
    };

    private final EnumMap<Direction, Integer> transformedCount = new EnumMap<>(Direction.class);

    @Override
    public boolean transform(final MutableQuadView mqv) {
        if (mqv.tag() == 1) {
            return true;
        }

        final Direction dir = mqv.lightFace();

        final int quadIndex = transformedCount.computeIfAbsent(dir, d -> 0);
        transformedCount.put(dir, quadIndex + 1);

        final int partIndex = getPartIndex(mqv, dir);

        final Data data = this.data[partIndex];

        final Pair<Float4, Float4> origUvs = getUvs(mqv, dir);

        // JMX defines models such that all quads appear twice
        // The first quad of a pair is used to render the base texture,
        // while the second quad is used to render the overlay.

        final Function<Pair<SpritesSource, BlockState>, Pair<Pair<Sprite, Optional<BlockState>>, OptionalInt>> findMaybeSpriteAndColor =
            pair -> pair.getFirst()
                .getSpriteAndColor(dir, quadIndex % pair.getFirst().getCount(dir), data.color)
                .mapFirst(s -> Pair.of(s, Optional.of(pair.getSecond())));

        if (quadIndex % 2 == 0) {
            if (!data.sprites.isPresent()) {
                return true; // no custom texture => stop transforming, show regular texture
            }

            final Float4 us = origUvs.getFirst();
            final Float4 vs = origUvs.getSecond();

            final Pair<Pair<Sprite, Optional<BlockState>>, OptionalInt> spriteAndColor = findMaybeSpriteAndColor.apply(data.sprites.get());

            final Sprite sprite = spriteAndColor.getFirst().getFirst();

            spriteAndColor.getFirst().getSecond().ifPresent(materialSource -> {
                final @Nullable RenderMaterial target = MaterialMap.get(materialSource).getMapped(sprite);

                if (target != null) {
                    mqv.material(target);
                }
            });

            applySpriteAndColor(mqv, sprite, spriteAndColor.getSecond(), us, vs);
            return true;
        } else {
            return data.overlay.match(
                overlay -> {
                    data.cachedOverlayColorApplier.apply(mqv);
                    return overlay.apply(mqv, origUvs.getFirst(), origUvs.getSecond(), dir);
                },
                () -> {
                    final Optional<Pair<Pair<Sprite, Optional<BlockState>>, OptionalInt>> maybeSpriteAndColor =
                        data.sprites.map(findMaybeSpriteAndColor);
                    final Float4 us = origUvs.getFirst();
                    final Float4 vs = origUvs.getSecond();

                    maybeSpriteAndColor.ifPresent(spriteAndColor -> {
                        final Sprite sprite = spriteAndColor.getFirst().getFirst();

                        spriteAndColor.getFirst().getSecond().ifPresent(materialSource -> {
                            final @Nullable RenderMaterial target = MaterialMap.get(materialSource).getMapped(sprite);

                            if (target != null) {
                                mqv.material(target);
                            }
                        });

                        applySpriteAndColor(mqv, sprite, spriteAndColor.getSecond(), us, vs);
                    });

                    return maybeSpriteAndColor.isPresent();
                }
            );

//            data.overlay.apply(mqv, origUvs.getFirst(), origUvs.getSecond(), dir);

//            final Optional<Overlay> maybeOverlay = data.overlay;
//
//            if (!maybeOverlay.isPresent()) {
//                maybeSpriteAndColor = data.sprites.map(findMaybeSpriteAndColor);
//
//                us = origUvs.getFirst();
//                vs = origUvs.getSecond();
//            } else {
//                final Overlay overlay = maybeOverlay.get();
//
//                overlay.textureSource().apply(mqv, us, vs, dir);
//
////                final Optional<TextureSource> maybeTextureSource = overlay.textureSource();
////                if (!maybeTextureSource.isPresent()) {
////                    maybeSpriteAndColor = Optional.empty();
////                } else {
////                    final TextureSource textureSource = maybeTextureSource.get();
////
////                    textureSource.apply(mqv, us, vs, dir);
////
////                    final Optional<TextureSource.Entry> maybeEntry = textureSource.entryFor(dir);
////
////                    if (!maybeEntry.isPresent()) {
////                        return false; // there is an overlay, but it doesn't have a texture for this direction
////                    }
////
////                    //noinspection deprecation
////                    maybeSpriteAndColor = maybeEntry.map(entry ->
////                        Pair.of(
////                            Pair.of(
////                                MinecraftClient.getInstance()
////                                    .getSpriteAtlas(SpriteAtlasTexture.BLOCK_ATLAS_TEXTURE)
////                                    .apply(entry.texture),
////                                entry.materialSource
////                            ),
////                            data.maybeCachedOverlayColor
////                        )
////                    );
////                }
//
//                final Optional<Map<Direction, Offsetters>> maybeSidedOffsetters = overlay.sidedOffsetters();
//                if (!maybeSidedOffsetters.isPresent()) {
//                    us = origUvs.getFirst();
//                    vs = origUvs.getSecond();
//                } else {
//                    final Map<Direction, Offsetters> sidedOffsetters = maybeSidedOffsetters.get();
//
//                    final Offsetters offsetters = sidedOffsetters.get(dir);
//                    if (offsetters == null) {
//                        us = origUvs.getFirst();
//                        vs = origUvs.getSecond();
//                    } else {
//                        us = offsetters.u.map(o -> o.offset(origUvs.getFirst())).orElseGet(origUvs::getFirst);
//                        vs = offsetters.v.map(o -> o.offset(origUvs.getSecond())).orElseGet(origUvs::getSecond);
//                    }
//                }
//            }
        }
    }

    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
    private static class Data {
        public final Optional<Pair<SpritesSource, BlockState>> sprites;
        public final Overlay overlay;
        public final ColorApplier cachedOverlayColorApplier;
        public final int color;

        public Data(final Optional<Pair<SpritesSource, BlockState>> sprites, final Overlay overlay, final OptionalInt maybeCachedOverlayColor, final int color) {
            this.sprites = sprites;
            this.overlay = overlay;
            this.cachedOverlayColorApplier = ColorApplier.ofOptional(maybeCachedOverlayColor);
            this.color = color;
        }
    }

    private final BlockState state;
    protected Data[] data;

    protected FrameTransform(final BlockRenderView brv, final BlockState state, final BlockPos pos, final Supplier<Random> randomSupplier) {
        this.state = state;

        if (!(state.getBlock() instanceof Frame)) {
            throw new IllegalArgumentException("Cannot apply frame transform to non-frame block " + state.getBlock() + ".");
        }

        //noinspection unchecked
        final Stream<Pair<Optional<BlockState>, Optional<Identifier>>> attachment =
            (Stream<Pair<Optional<BlockState>, Optional<Identifier>>>) ((RenderAttachedBlockView) brv).getBlockEntityRenderAttachment(pos);

        //noinspection ConstantConditions
        data = attachment.map(pair -> {
            final Optional<BlockState> maybeBaseState = pair.getFirst();

            final int color;
            final Optional<Pair<SpritesSource, BlockState>> sprites;

            if (maybeBaseState.isPresent()) {
                final BlockState baseState = maybeBaseState.get();
                color = Optional.ofNullable(ColorProviderRegistry.BLOCK.get(baseState.getBlock()))
                    .map(prov -> prov.getColor(baseState, brv, pos, 1) | 0xFF000000)
                    .orElse(0xFFFFFFFF);
                sprites = Optional.of(Pair.of(
                    new SpritesSource(baseState, MinecraftClient.getInstance().getBlockRenderManager().getModel(baseState), randomSupplier.get()),
                    baseState
                ));
            } else {
                color = 0xFFFFFFFF;
                sprites = Optional.empty();
            }

            final Overlay overlay = pair.getSecond().map(CLIENT_OVERLAYS::getOverlayFor).orElse(Overlay.NONE);

            final OptionalInt cachedOverlayColor =
                flatMapToInt(
                    overlay.coloredLike(),
                    coloredLike -> mapToInt(
                        Optional.ofNullable(ColorProviderRegistry.BLOCK.get(coloredLike.colorSource().getBlock())),
                        prov -> prov.getColor(coloredLike.colorSource(), brv, pos, 1) | 0xFF000000
                    )
                );

            return new Data(sprites, overlay, cachedOverlayColor, color);
        }).toArray(Data[]::new);
    }

    protected int getPartIndex(final MutableQuadView mqv, final Direction dir) {
        return ((FrameSlotInfo) state.getBlock()).getRelativeSlotAt(
            state,
            new Vec3d(
                calcCenter(mqv::x),
                calcCenter(mqv::y),
                calcCenter(mqv::z)
            ),
            dir
        );
    }

    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
    protected void applySpriteAndColor(final MutableQuadView mqv, final Sprite sprite, final OptionalInt maybeColor, final Float4 us, final Float4 vs) {
        mqv.sprite(0, 0, MathHelper.lerp(us.a, sprite.getMinU(), sprite.getMaxU()), MathHelper.lerp(vs.a, sprite.getMinV(), sprite.getMaxV()));
        mqv.sprite(1, 0, MathHelper.lerp(us.b, sprite.getMinU(), sprite.getMaxU()), MathHelper.lerp(vs.b, sprite.getMinV(), sprite.getMaxV()));
        mqv.sprite(2, 0, MathHelper.lerp(us.c, sprite.getMinU(), sprite.getMaxU()), MathHelper.lerp(vs.c, sprite.getMinV(), sprite.getMaxV()));
        mqv.sprite(3, 0, MathHelper.lerp(us.d, sprite.getMinU(), sprite.getMaxU()), MathHelper.lerp(vs.d, sprite.getMinV(), sprite.getMaxV()));

        if (maybeColor.isPresent()) {
            final int color = maybeColor.getAsInt();

            mqv.spriteColor(0, color, color, color, color);
        }
    }

    protected Pair<Float4, Float4> getUvs(final MutableQuadView mqv, final Direction dir) {
        final IntStream us = IntStream.rangeClosed(0, 3);
        final IntStream vs = IntStream.rangeClosed(0, 3);
        switch (dir) {
        case DOWN:
            return Pair.of(
                Float4.fromIterator(us.mapToDouble(i -> MathHelper.clamp(mqv.x(i), 0, 1)).iterator()),
                Float4.fromIterator(vs.mapToDouble(i -> 1 - MathHelper.clamp(mqv.z(i), 0, 1)).iterator())
            );
        case UP:
            return Pair.of(
                Float4.fromIterator(us.mapToDouble(i -> MathHelper.clamp(mqv.x(i), 0, 1)).iterator()),
                Float4.fromIterator(vs.mapToDouble(i -> MathHelper.clamp(mqv.z(i), 0, 1)).iterator())
            );
        case NORTH:
            return Pair.of(
                Float4.fromIterator(us.mapToDouble(i -> 1 - MathHelper.clamp(mqv.x(i), 0f, 1f)).iterator()),
                Float4.fromIterator(vs.mapToDouble(i -> 1 - MathHelper.clamp(mqv.y(i), 0f, 1f)).iterator())
            );
        case SOUTH:
            return Pair.of(
                Float4.fromIterator(us.mapToDouble(i -> MathHelper.clamp(mqv.x(i), 0f, 1f)).iterator()),
                Float4.fromIterator(vs.mapToDouble(i -> 1 - MathHelper.clamp(mqv.y(i), 0f, 1f)).iterator())
            );
        case EAST:
            return Pair.of(
                Float4.fromIterator(us.mapToDouble(i -> 1 - MathHelper.clamp(mqv.z(i), 0f, 1f)).iterator()),
                Float4.fromIterator(vs.mapToDouble(i -> 1 - MathHelper.clamp(mqv.y(i), 0f, 1f)).iterator())
            );
        case WEST:
            return Pair.of(
                Float4.fromIterator(us.mapToDouble(i -> MathHelper.clamp(mqv.z(i), 0f, 1f)).iterator()),
                Float4.fromIterator(vs.mapToDouble(i -> 1 - MathHelper.clamp(mqv.y(i), 0f, 1f)).iterator())
            );
        default:
            throw new IllegalArgumentException("Invalid direction: " + dir);
        }
    }
}
